src/main/java/org/jenkinsci/plugins/scriptsecurity/sandbox/groovy/SandboxInterceptor.java

```diff
@@ -78,8 +78,13 @@
             selfArgs[0] = receiver;
             System.arraycopy(args, 0, selfArgs, 1, args.length);
             for (Class<?> dgmClass : DGM_CLASSES) {
-                if (GroovyCallSiteSelector.staticMethod(dgmClass, method, selfArgs) != null) {
-                    return onStaticCall(invoker, dgmClass, method, selfArgs);
+                Method dgmMethod = GroovyCallSiteSelector.staticMethod(dgmClass, method, selfArgs);
+                if (dgmMethod != null) {
+                    if (whitelist.permitsStaticMethod(dgmMethod, selfArgs)) {
+                        return super.onMethodCall(invoker, receiver, method, args);
```

Comment:
Welcome to Groovyland. https://github.com/jenkinsci/workflow-plugin/pull/372 demonstrates what was wrong and what now works with this PR. The user invokes `each` on some `Object` (such as a `List`), passing a `Closure` as argument.

Normally Groovy translates this call into `static DefaultGroovyMethods.each(Object, Closure)`, like any other “GDK” methods which are injected into types that did not originally support the method. So we check the whitelist on that. Prior to this PR, we then delegate back to Groovy’s call site logic with the `DefaultGroovyMethods` replacement, which normally works.

But after `groovy-cps` hacks at the method dispatch tables so that such an `each` call now delegates to `CpsDefaultGroovyMethods.each(Object, Closure)`, when running in the sandbox we were still forwarding to the unpatched method.

So this fix means that we continue to check the same whitelist entry, but we now go back to Groovy’s own official lookup table if the call is approved, just like we do for all other kinds of method calls, so the patch to `CpsDefaultGroovyMethods` takes effect.